### PR TITLE
Add datafusion-substrait crate

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4536

# Rationale for this change

Move development of datafusion-substrait into main DataFusion repo.

# What changes are included in this PR?

- Add new `datafusion/substrait` folder, containing contents of `datafusion-substrait` repo, preserving history.
- This crate is not part of the workspace yet and depends on an older version of DataFusion. A follow on PR (https://github.com/apache/arrow-datafusion/pull/4879) will integrate the crate into the workspace and update the release scripts.

# Are these changes tested?

Tests are included.

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by andygrove** [comment]
@waynexia @nseekhao @JanKaul I need to go through the ASF IP clearance process to have this code moved into this repo. Could you please follow the instructions at https://www.apache.org/licenses/contributor-agreements.html to submit an ICLA (Individual Contributor Licence Agreement) if you do not already have one on file? Thanks in advance for helping with this!

**Comment by andygrove** [comment]
Thanks @nseekhao and @JanKaul for filing ICLAs.

@waynexia Will you also be able to file an ICLA?

File: datafusion/substrait/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# under the License.
+
+[package]
+name = "datafusion-substrait"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+async-recursion = "1.0"
+datafusion = "13.0"
```
[the remark this answers is not in the document]
Comment (andygrove) [reply datafusion/substrait/Cargo.toml:25]: This PR is to accept the donation of the current codebase, and then we will update in future PRs to integrate fully with the project and use the current DataFusion version.

**Comment by waynexia** [comment]
Sorry for the delay 🥲 I just got feedback from the secretary. I'll fix the issue and resubmit it today.

**Comment by andygrove** [comment]
@waynexia @JanKaul @nseekhao @Dandandan Thanks everyone for filing ICLAs. As part of the IP clearance process, I have to remind active committers that they are responsible for ensuring that a Corporate CLA is recorded if such is required to authorize their contributions under their individual CLA. 

I have filled out the necessary "paperwork" and will file this in the next day or two and start a vote on accepting this donation. Thanks for your patience.

**Comment by waynexia** [comment]
Thanks for handling this @andygrove

**Comment by andygrove** [comment]
Vote thread: https://lists.apache.org/thread/go28mx024wxnjrpznmtfzd3v5dt49qzp

**Comment by andygrove** [comment]
IP clearance form: https://incubator.apache.org/ip-clearance/arrow-datafusion-substrait.html

**Comment by andygrove** [comment]
The Apache Arrow PMC has voted to accept this contribution:

https://lists.apache.org/thread/qcbdt8y2vkwvwkjmmljn4jzvggzb2fkk

A lazy consensus incubator vote has now started:

https://lists.apache.org/thread/m5q4qxr32xzhotljv6z7mg3dofl46rv6

**Comment by andygrove** [comment]
The incubator vote passed. Given that the PMC has already voted to accept this donation I plan on merging this PR later today.

File: datafusion/substrait/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andygrove) [reply datafusion/substrait/Cargo.toml:25]: Here is the follow-on PR for this: https://github.com/apache/arrow-datafusion/pull/4879

**Review by alamb — APPROVED** [review]
It is great to see this finally land @andygrove  -- thank you!

**Pull request merged**